### ISSUE TITLE
Deposit refactor

### ISSUE THREAD

**Issue by JosephMcArthur** [oaworks/discussion#1829]
We've had a refactor of the deposit route on the list for a while. A few recent issues made us discuss:

* https://github.com/oaworks/discussion/issues/1826
* https://github.com/oaworks/discussion/issues/1820
* https://github.com/oaworks/discussion/issues/1830
* https://github.com/oaworks/discussion/issues/1832
* Bad external scripts generating 100k or so emails.

In discussion we agreed that the work didn't have to be major but needed planning and wasn't a priority right now. So, just setting this up to track / discuss.

cc @markmacgillivray

**Comment by JosephMcArthur** [comment]
Discussed again due to #1830, #1832. 

@markmacgillivray  would like to see us move on this. 

The path is that we do a real refactor and move to the new codebase but without taking into account possible but not confirmed new requirements. Then, if the test pass, we're good. I have some fears our test coverage won't be good enough but, somewhere we'll have the original test files and results so technically we should be able to recreate that fairly easily.

**Comment by JosephMcArthur** [comment]
Status: Code in new system is done. Needs testing. 

- [ ] @markmacgillivray switch dev to this new version. Review Ghost insepector for obvious issues & address. 
- [ ] @markmacgillivray let Joe know when he should test.

**Comment by markmacgillivray** [comment]
Keep in mind solution to #1834 should come from switching to new deposit too - by having deposit use the same permissions at all steps, should not end up with occasional mismatched licences.

**Comment by markmacgillivray** [comment]
Dev deposit has been switched over, but I am updating the new system to be able to run legacy filetype conversions, as that was the first issue I hit with an old doc file. Once that is done I will report if ready for testing deposit on dev.

**Comment by JosephMcArthur** [comment]
Thanks, can you give me an update tomorrow on the likelyhood of this being testable by me (i.e all syp deposit tests passing and no known issues on your end). I have a two dozen files that need depositing manually due to all the bugs atm, and it would be a two birds in one if I could use those to test this & deposit them! But, can't wait on them forever.

**Comment by markmacgillivray** [comment]
This is being delayed by needing to get the text extractors working - they're finicky, and won't work on the new machine. I will need to write some alternatives. However that's taking too long given the need to do another Gates report, so I will switch to doing some Gates stuff then come back to this.

**Comment by JosephMcArthur** [comment]
Thanks, really appreciate the update.

**Comment by markmacgillivray** [comment]
Unfortunately I have not yet been able to get the text extractors to work with our new structure. So work on this is continuing. We can't finish deposit until we can extract text, because deposit has to be able to extract the text from whatever file type it is given and check if it is the right file.

**Comment by markmacgillivray** [comment]
Our most common file types of doc, docx, and pdf can now have text extracted in the new system. html and xml are sometimes necessary too, but they should be easy to add. I should be able to get the deposit refactor connected up to the new converters tomorrow, then ready to test deposit on dev tomorrow / Tuesday.

**Comment by JosephMcArthur** [comment]
Fwiw, in my memory I've never seen a html or xml file come through. Happy
to skip those.

If you could spend some time testing that would be great. I'll likely be
traveling tomorrow so I'll be slower than usual. All tests should pass at
least.

**Comment by JosephMcArthur** [comment]
Just dug out the testing sheet we used to validate the API the first time around:

https://docs.google.com/spreadsheets/d/15roBCasoxe_DXtty2iN-QZU6uz4urqn5hTDjNVIe8_c/edit#gid=0

You'll find test files linked, and their manually validated correct version in column F.

If you could test against those that would be fab too and help us move this along with high confidence quickly.

**Comment by markmacgillivray** [comment]
I have connected dev deposit to the new API dev deposit. I've tested that files can be uploaded, text extracted, files accepted, and deposited into zenodo (sandbox). I now need to update the mail templates management as the new API reads them from a sheet rather than the repo, but that sheet isn't populated yet. I'll stick the necessary deposit templates into it tomorrow, then will run through the wider testing of deposits.

I didn't do html or xml extraction yet, as you've said we can skip them.

**Comment by markmacgillivray** [comment]
Work to integrate templates from sheets is done, a complete deposit can now be done on dev, however the confirmation email doesn't get sent even though the templates are read and the deposit returns successfully. Will do more work tomorrow to find out what is happening to the emails.

**Comment by markmacgillivray** [comment]
On dev it is now fully working with emails being sent through too. Now I can move on to testing with the sheet you linked above.

**Comment by markmacgillivray** [comment]
We should make an issue to transfer more templates from the old website repo into the sheet they now get imported from. So far I only did the instantill and deposit ones, not sure which others are still going to be relevant for the future of the system. Which repo would you want such an issue to be in?

**Comment by JosephMcArthur** [comment]
Discussion please

**Comment by markmacgillivray** [comment]
OK done:

https://github.com/oaworks/discussion/issues/1841

**Comment by markmacgillivray** [comment]
Extracting texts works but there was an issue with matching titles, turned out to be due to formatting differences in the new text extraction software. This has been fixed now and titles match.

Random tests were made from the test sheet throughout the last week or so. Now I've added a column and started noting "yes" when they work on the new API.

However, so far this only works passing the files direct to the background server - which is where the text extraction has to run. I need to get cloudflare to hand off the request to the background machine when a file comes in with it. It already hands off to background on error or when set to do so, but at the moment it's not passing the file. So need to get that working now, so that the main frontend (cloudflare worker) URL can be used but still pass off in the background. Otherwise we'd have to explicitly send file uploads to the bg address which is messy and fragile.

**Comment by markmacgillivray** [comment]
I think this is tested sufficiently to show that the new API returns the same results as the old. I've also created a new embed repo and updated that embed to send deposits direct to the background server. I will update the old dev website to use the new embed code, then some user testing of the dev shareyourpaper gates page would be a good idea (user can be whichever OA.Works person knows what to expect from the page progressions).

**Comment by JosephMcArthur** [comment]
It doesn't look like all tests are passing. Can you look into that first
please?

**Comment by markmacgillivray** [comment]
I went through the SYP tests and either passed them or found legit reasons why they failed, was going to document here, but then found that the zenodo test API is down, so can't recreate the same as before. This may be why more are failing now than were failing earlier in the week. On Monday next week I'll check again and if zenodo sandbox is working again I can write up the reasons.

I also spent some time working on the separated embed code, because I'd started that already and considered testing it along with the deposit refactor, since it's going to need basically all the same tests done again. Unfortunately switching to use the new separate embed on the old site would take more fiddling of other parts that may not be worth it given that some of that may not be kept long term. So I'll finish this deposit refactor issue separately next week, then will come back to how it works with the separated embed code early in new year.

**Comment by markmacgillivray** [comment]
On SYP tests:

* admin aam override fails because the paper really is already freely available, whereas the test expects it not to be.

* dark deposit library default seems to fail because it looks for content in an email but can't find it. But the email I receive a copy of appears to have the content in it. I don't know enough about how you set up these tests to understand what it is doing here - and this may be relevant to some other test failures. @JosephMcArthur can you take a look or let me know what I'm supposed to do about this?

I will continue with other test write-ups in the next comment.

**Comment by markmacgillivray** [comment]
* deposit email validation fails because the email does not get rejected. I think we agreed to stop using the email validation though, after a policy change at mailgun. Are you still expecting this to work?

**Comment by markmacgillivray** [comment]
* embargo length display fails because the article in question was in embargo but that embargo ended in June 2021

**Comment by markmacgillivray** [comment]
* pub pdf upload fails because it expects the file to get sent to zenodo, but as far as I can tell the system correctly identifies it as likely a publisher pdf. Reasons being it is a PDF, and it contains two regex matches for things that indicate it is a publisher PDF. Do you expect this to be different?

**Comment by markmacgillivray** [comment]
* redeposit (library default) appears to fail because it is another one that attempts to look into an email. Any more info you may have about this would be helpful

**Comment by markmacgillivray** [comment]
* redeposit (syp homepage) fails because it expects a different URL to the file than the one the system currently gives. However the one the system gives does work. Does it need to be the one you've tested for, or can this test just be changed?

**Comment by markmacgillivray** [comment]
That's all the SYP tests tested and retested, and avoiding zenodo sandbox issues that were present before christmas.

**Comment by JosephMcArthur** [comment]
> admin aam override fails because the paper really is already freely available, whereas the test expects it not to be.

Passing.

Fixed by changing `10.1523/JNEUROSCI.2415-19.2020` to `10.1016/j.aap.2021.106546`. 10.1016/j.aap.2021.106546 is in `0001-4575` which has a three year embargo so hoping this doesn't change soon.

> dark deposit library default seems to fail because it looks for content in an email but can't find it. But the email I receive a copy of appears to have the content in it. I don't know enough about how you set up these tests to understand what it is doing here - and this may be relevant to some other test failures. @JosephMcArthur can you take a look or let me know what I'm supposed to do about this?

Passing.

I have an email filter set up in my inbox such data emails for this test get forwarded to ghost. I do this because the test embed is set up under my address and isn't only used for ghostinspector so it's useful to get these to my email. When you changed deposit@openaccessbutton to deposit@oa.works that filter broke.

I have updated the filter, and now this works.

> deposit email validation fails because the email does not get rejected. I think we agreed to stop using the email validation though, after a policy change at mailgun. Are you still expecting this to work?

Test deleted

> embargo length display fails because the article in question was in embargo but that embargo ended in June 2021

Passing. Fixed as in first point.

> licence display (no publisher licence) fails because the article is already freely available

Passing. Fixed by changing `10.1523/JNEUROSCI.2415-19.2020` to `10.1093/adaptation/apab001` which has a long embargo and so hopefully won't change soon.

> pub pdf upload fails because it expects the file to get sent to zenodo, but as far as I can tell the system correctly identifies it as likely a publisher pdf. Reasons being it is a PDF, and it contains two regex matches for things that indicate it is a publisher PDF. Do you expect this to be different?

Failing correctly & needs fixing. The permissions API says the upload can be a `publishedVersion`, so, the deposit should be allowed. Can you look into why it isn't being allowed?

> redeposit (library default) appears to fail because it is another one that attempts to look into an email. Any more info you may have about this would be helpful

Passing. Fixed as above.

> redeposit (syp homepage) fails because it expects a different URL to the file than the one the system currently gives. However the one the system gives does work. Does it need to be the one you've tested for, or can this test just be changed?

Passing. Was fine to changed, and I've changed.

**Comment by JosephMcArthur** [comment]
@markmacgillivray, it looks like the deposit emails are being signed by `mmail.cottagelabs.com`. Can you share why, as discussed we're trying to separate our infra from CL so keen to know why this is now happening.

**Comment by JosephMcArthur** [comment]
An update here first thing next week would be fab if you can.

**Comment by markmacgillivray** [comment]
The email signing has been fixed, is now mg.openaccessbutton.org. It was still just running on the defaults which were the CL settings, but the OAB settings were in there too so I just turned them on.

The embargo length display test was failing because the new article you gave it is in a different journal, so the test condition needed changed. I changed it to match the new journal name and it works now. There was also a discrepancy in date format so I altered that too.

The pub pdf upload one was failing because the archivable algorithm did not account for permissions saying a published version is allowable. So I have added to allow that if permissions are known for the file (which they should be for a deposit). I have given it an "archivable_reason" of "The file given is a Publisher PDF, and permissions indicates that publisher PDFs are allowed".

**Comment by markmacgillivray** [comment]
If you are happy with all the above, and all the SYP tests passing, this would appear to be ready to go onto live. Let me know.

**Comment by markmacgillivray** [comment]
One thing to note about making this live - so far the deposit configs still come from the old system, and still will after this goes live. Once we move the old accounts to the new API, then configs will also be from the new API. Before doing that though, presumably some work has to be done on the config creation page which is still on the old website. This doesn't delay me on the API side of things, just mentioning it. However it will also mean that new deposits will be recorded in the new API whereas old deposits will still be in the old API, and all the functionality to view past deposits is also still tied to old user accounts on the old API. We will be recording new deposits, so we're not losing track of them, but nobody viewing their deposit history would see any new deposits after this change, up until we complete the total merge of old API with new. Or we could do some specific work just to move the old deposits before finally moving the accounts/configs. It seems to me at the moment our priority is still Gates use case, so pushing this live now still seems like the most useful thing to do. Then we can decide the priority of the deposit histories and configs.

Another related thing, which we already have a separate issue for anyway, is the new embed. At the time we do that we should also change the old website to use the new API routes, as the new embed will also use them. For now the old embed still contacts the old API, and the old API is now just getting the answers from the new API.

**Comment by JosephMcArthur** [comment]
> If you are happy with all the above, and all the SYP tests passing, this would appear to be ready to go onto live. Let me know.

Will hopefully review tomorrow and do some sanity checks outside of automated tests.

> We will be recording new deposits, so we're not losing track of them, but nobody viewing their deposit history would see any new deposits after this change, up until we complete the total merge of old API with new.

This seems less than ideal and I wasn't aware of this until now.

> Or we could do some specific work just to move the old deposits before finally moving the accounts/configs

What does this entail, how long will it take, and will we have to do it eventually anyway to merge everything?
Can we provide two endpoints, one before & one after the shift, so folks can access their complete deposit history until the two can be merged?

**Comment by JosephMcArthur** [comment]
Note: Some work is being done via: `Providing Embed community information`

- [ ] update deposit emails to include `community`
- [ ] Consider `record every single deposit` vs `iterate a "duplicate" counter`

**Comment by markmacgillivray** [comment]
We will already have two endpoints for deposits where someone could look up old ones on the old system and new ones on the new system, yes. If you want them to be able to see all old ones on the new system, now or in future, then yeah we have to move them across at some point - when that point is, depends how important you think that is for users.

Moving them across would not be too hard, would just have to write a script to change them to meet the new style as much as possible. But this is only worth doing if you do want all old records of deposits to persist in the new system. Whether or not that is useful would depend on what you think the users want to do with that historical data, if anything. 

Note also that if we are going to record every deposit rather than duplicate, we should definitely do that now before making it live rather than later. It's not hard to make that change, and the many copies isn't too many for the system to handle, it was more that it was more complicated for what you thought the users might want to do with that data, I think. E.g. they would have a record of every deposit attempt, but they wouldn't be related to each other, so some additional work would need to be done by the users to see that a particular set of deposits may all have been about the same thing (e.g. grouping by DOI once they export it to a sheet or whatever system they want to review it in). I think this is probably a good idea to do now though, because changing to duplicates later if that becomes something specifically asked for by any user is easier than having duplicates now and trying to change to uniques later. So if you can confirm this, I'll make this change before making it live.

At:

https://bg.beta.oa.works/deposits

You can see deposit records in the new (dev) system. If you search the page for "community" you will see one that has a user config present and has the community value in a specifically named field too. It's a list at the moment, but would change that to just a string if we get rid of counting duplicates and do uniques instead. You can see some duplicate counts in there too, unsurprisingly for our test records the number can be in the 100s already.

**Comment by JosephMcArthur** [comment]
Did a few minutes of sniff tests today and found an issue.

Attempted to deposit `10.1038/nature20796` (using [Nature 2016 accepted open access (1).pdf](https://github.com/oaworks/discussion/files/7840657/Nature.2016.accepted.open.access.1.pdf)) at `https://dev.shareyourpaper.org/funders/gates?doi=undefined&email=eran.elinav@weizmann.ac.il`. However, I never got the review emails (I tried this several times). I'm a little confused, as the test for this workflow is passing: https://app.ghostinspector.com/tests/5f23df3c5e12c014aeaf08d5. But, I could replicate the issue with `10.1038/s41586-019-1443-5`.

Separately, and I'm not sure if this really matters or is related, but when I hit `do another` there the URL i'm taken too is `https://shareyourpaper.org/funders/gates?doi=undefined&email=eran.elinav@weizmann.ac.il`. I think it's fine to leave the email there but having `doi=undefined` feels inelegant.

**Comment by JosephMcArthur** [comment]
I think we should unduplicate in the API. Looking now it is confusing to know what things are / aren't duplicated there now, and what is hidden by the duplication. I think it's important to have a clear source of truth, and then, as required filter from there (e.g in a bulk import use case)

**Comment by JosephMcArthur** [comment]
I just tried another deposit, https://dev.shareyourpaper.org/10.1016/j.jmapro.2019.05.029 with [manuscript_accepted.pdf](https://github.com/oaworks/discussion/files/7840981/manuscript_accepted.pdf).

This succeeded on live yesterday, as expected. On dev it was flagged as the wrong version. Again, [the test for this workflow is passing](https://app.ghostinspector.com/tests/5f23dcc15e12c014aeaefb71).

**Comment by markmacgillivray** [comment]
I have unduplicated the deposits API. I haven't looked at doi=undefined yet.

For the ongoing testing of this and to add to our discussion of how long it has taken, whilst I agreed yesterday this is partly down to my low productivity over the last three or four weeks, I also want to use this as an example to show why something that I may initially say would be quite quick ends up taking quite a long time. Not to blame, but to help us improve process.

I've already explained that I had to spend significant extra time going through the 9 or 10 ghost inspector tests that were failing, and that was time I hadn't accounted for. As you also acknowledged, it took you a few hours to dig through just one of them. The reason I hadn't accounted for this is that I didn't consider those tests a source of truth, they are out of date, you'd previously indicated awareness and acceptance of at least one of them being wrong, and indeed, the answers in almost all cases after I reviewed were that my code was correct but the tests had become obsolete, or something else about the process expectations had changed. I have no problem with doing more extensive testing in future if that is desired, and if so I should include that in the time estimates to complete a piece of work. I saw these more as UI/user tests, and never wrote them nor maintained them, so didn't see them as my responsibility. Again, to clarify, I'm totally happy if they should be within my role, but that's not been clear in the past. It may seem that testing should be clearly up to me, and that I should check these tests before handing stuff to you for confirmation, but as mentioned at the start, I instead checked my work against what I expected the results to be, and it turns out I was more accurate than the tests. So I couldn't have known that there was more work I should have done.

In relation to your recent sniff tests, we see this problem arise again. It seems you have different expectations about what it would mean for the system to be successful, and I don't think it is possible for me to have known that. In the case of you not receiving review emails, I've spent a fair bit of time investigating that issue now, going through new and old code, and then going through old issues, and it seems from what I can find that you previously instructed me to turn those off due to the spam you were receiving. You've actually linked to that issue further up in the comments here. So, I do want to ensure I hand good work to you rather than waste your time, but I don't think I could have known that this should be doing something different. The new code I wrote appeared to match the old behaviour, and the old behaviour matched your instructions. The ghost inspector test passes because that test submits a file, and previous instructions allowed for review emails to be sent in the case where files are submitted. Again, this is what the old code did, and what the new code was written to do, and as you say, in this case it even passed the test. So unfortunately there's no way I could have known you expected something different, nor that I should have spent the extra day of work I've now spent checking for that. Again, I'm happy for that to be a responsibility of my role, but it would entail greater time estimates if I need to account for the possibility that you actually wanted the system to not only be refactored but to behave differently to how it did before, and to how it does in successful testing.

For the "manuscript_accepted" case, as far as I can tell, it correctly failed because the file name contains "accepted.pdf" which is listed as an indicator string to pick up. The ghost inspector test passes because although the file content that it uses is the same, the file NAME is NOT the same. So here again we have an incorrect test, that I could not have known should have been doing something different, and all my review of code indicated that I had completed the work satisfactorily. There is also an additional complexity here, in that during this deposit refactor we HAVE made a change beyond refactoring based on a comment you made, which is that we do now allow PDFs even if they appear to be the published version IF permissions says published versions are allowed. Previously, if it was a PDF we just didn't allow it. So now we have a new part here allowing for PDFs that are published versions, and I had already documented that above, and asked if you were happy with that. It turns out, as far as I can discern from your most recent accepted manuscript test, that we also need further enhancements here to also accept PDFs that are acceptedVersion if permissions states those are accepted. So I've added that too, now. I've also added another catch for if we appear to have a submittedVersion, and I've assumed that would always be acceptable. Let me know if that's not the case. Note, I again couldn't have known in advance that you would want these enhancements done when I estimated how long this would take, and indeed it isn't finished because I can't complete this without knowing if these are how you want them or not.

I think I'd have to say at this point that I feel I was justified in saying nearly four weeks ago that I believed my work here was done sufficiently to meet requirements to replace the current live system with the new refactored code. Everything since then appears to be down to unplanned changes, obsolete tests, and holiday period time off.

To estimate the discrepancy between how long it has taken and what we originally expected - yesterday we discussed for about 3 hours, which I'll guess was about half on this topic. Since then I've worked about another 8 investigating these latest points. So say round up to 10 hours of additional work, just this week. If we extrapolate that back to the start of December, that's at least 7 days of unexpected work that went into this. That doesn't mean it was a waste, but I hope it does help to ease your confusion of why you feel you've been waiting so long.

What could we do to improve this in future?

* define which role is responsible for creating/maintaining/running tests
* (and we're already aware of the problem that for the time being those roles may have to fall on the "wrong" person until we hire more people to fill them - but it probably still helps to define them)
* include time to review and run tests in every estimate (or have a job role with time dedicated to maintaining and running them as a stand-alone process)
* don't allow obfuscated scope creep during a time-sensitive milestone, e.g. where a refactor ends up introducing new development opportunities
* ensure that newly discovered developments that would be worthwhile are only introduced to scope by being added to the queue of work as a specific new issue, to make it more visible that it is EXTRA work we shouldn't feel bad about
* (and as already agreed yesterday, I need to manage my context-switching better in light of unavoidable new lifestyle impacts that I need to ensure don't interfere with work so much)

Hope this, helps, happy to discuss further.

**Comment by markmacgillivray** [comment]
To list what should be confirmed/altered to complete this:

* is the change for accepting PDFs OK? I've made it so if the uploaded file is a PDF and version is acceptable by permissions, the archivable_reason is "The file given is the X version, and permissions indicates that is allowed", or if it is not allowed it will say "The file given is the X version, and only Y version is allowed". If the version appears to be a submitted version then it is always acceptable and would say "The file given is a submitted version, which are allowed"
* Are the current tests what should be expected? If not, can we review the tests and set them to what we actually need them to be, then I'll go through them all again to ensure they pass.
* I seem to get review emails, and the test passes for review emails. Can we clarify what it is that you're expecting but not seeing, or can we confirm that it is doing what it should do already?
* Is it OK to leave old deposit history records on the old system? Or if not, to move them at a later date? This shouldn't block Gates work, but may start looking odd to old users who notice they don't see more recent deposits when the check the deposits API (which they will be doing against the old system). If we prioritise this to be done soon, they may never notice.
* Is the format of the new deposits, which have now also been set to not count duplicates but just always create new records, acceptable now?

**Comment by markmacgillivray** [comment]
Note: in discussion we have found the specific issue with 10.1038/nature20796 which is that mailgun is returning a 413 request entity too large, which either is due to a change at mailgun or due to some change with how we are forwarding the attached file, AFTER the user has confirmed the file should have been correct, or where the deposit was successful and the success deposit email should have sent.

**Comment by markmacgillivray** [comment]
From discussion: check if the "new" functionality where deposit checks the file version against what permissions says is allowed is the same as what the old permissions/file system used to do. (New permissions doesn't do that part because it was removed during the permissions refactor)

**Comment by markmacgillivray** [comment]
I have taken the old algorithm from permissions that matched the version against what the file version was, and have added it into the new deposit algorithm.

Old code can be seen at:
https://github.com/oaworks/API/blob/master/noddy/service/v2/permissions.coffee#L422

New code can be seen at:
https://github.com/oaworks/paradigm/blob/830dc628893096ec80e1115269aabc37c7a84a65/worker/src/svc/oaworks/deposit.coffee#L314

Please review them and let me know if this is sufficient enough to make you confident that they behave the same, and as you expect.

**Comment by JosephMcArthur** [comment]
Thanks! Once you've looked into Mailgun I'll take another look at
everything here.

Joseph McArthur
Assistant Director: Right to Research Coalition <http://righttoresearch.org>
Director / Founder: Open Access Button <http://openaccessbutton.org>
Twitter: @Mcarthur_Joe <https://twitter.com/Mcarthur_Joe>
Skype: joseph_mcarthur

**Comment by markmacgillivray** [comment]
Sending to mailgun is partially fixed now in that the emails are not just disappearing, however now they don't have the attachments on them. Continuing to work on this.

**Comment by markmacgillivray** [comment]
Mailgun is now receiving and forwarding on the attached files for review deposits. Code is committed into develop:
https://github.com/oaworks/paradigm/commit/b7d2b2159b1c19df6755d2c2776f798a429d3bad

(testing as the commit msg means testing a change to the workflow to start adding more commit msg information, will improve next time)

**Comment by markmacgillivray** [comment]
Yes I'm doing other work on dev that will have an effect, I thought that was OK because you wouldn't get to looking at this this week. I can change it back if you're wanting to check these tests this week though.

**Comment by JosephMcArthur** [comment]
On Friday or over the weekend there is a reasonable chance of me being able
to have a look (fingers crossed). But, if you've got it changed for a
reason then use your judgement on if it makes sense to change back or if it
can be changed back when you're not working, for instance.

Joseph McArthur
Assistant Director: Right to Research Coalition <http://righttoresearch.org>
Director / Founder: Open Access Button <http://openaccessbutton.org>
Twitter: @Mcarthur_Joe <https://twitter.com/Mcarthur_Joe>
Skype: joseph_mcarthur

**Comment by markmacgillivray** [comment]
It turns out the failures are due to permissions changes, not other dev/deposit work. I've already put some questions on the permissions issue that I need answered before that can be finished. Of course, deposit uses permissions first, so you can't get to the expected part of deposits if the initial permissions check comes back different to what is expected. So if you have time on Friday / over the weekend, please have a look at permissions stuff first then I can finish the work to get the imports to work from the new sheets, then permissions and deposit will go back to normal working. At the moment permissions isn't able to import all the data correctly, so that's why we need to finish that first now.

**Comment by markmacgillivray** [comment]
See #1590 for related progress

**Comment by JosephMcArthur** [comment]
Looks like a couple of tests are failing. In both cases, there are no strings matching where there should be some matching strings (e.g $100). Can you look into that? Would be fab if that was possible for you early on Monday so I can test the rest of the day before I take a few days off.

**Comment by markmacgillivray** [comment]
Yeah I saw those on Friday, didn't have time to finish them though. But I've done it now. It was due to the change to google sheets, turns out the one place that did rely on the changes I had manually added was the File checking rules sheets, which has some oddities to the column headers (e.g. "what it Indicates" has a random capital I in it, whereas everything else was lowercase and had spaces). I've added a catch for that and checked for anywhere else that may have relied on those columns being cleaned.

However when it comes to the matches, it turns out that $100 regex doesn't work anyway. I got the test to pass by matching the "Downloaded at" string though, and catching the error of the failing regex. I've also added a strings_errored to the API output so you can see if some are causing errors. If you do the usual test paper upload at:

https://dev.shareyourpaper.org/10.1063/1.4937158

using the paper specified on the ghostinspector SYP test "pub pdf upload" (one of the ones that was failing) you can see in the devtools that the API response now includes "strings_errored" and it shows that regex with the error:

SyntaxError: Invalid regular expression: /^((USD|EUR|€|\\$)\\s?\\d{1,2}([\\.\\,]\\d{2})?)$|^(\\d{2}([\\.\\,]\\d{1,2})?\\s?(USD|EUR|€|\\$))$/: Invalid escape

So I think that is actually a separate issue, and not one that blocks this issue or further testing of the deposit refactor. Now that I've added catches, the SYP tests pass.

**Comment by markmacgillivray** [comment]
So you should be free to test on Monday before you leave.

**Comment by JosephMcArthur** [comment]
I've tested a dozen or so recent / old deposits, and everything went as expected. Given all tests are passing I'd say let's get this released as soon as possible.

I wouldn't be shocked if there were a couple of issues still –– but I'm confident things are better. With that in mind, and because I'm going away, @markmacgillivray can you notify @swipperman when this goes live too? @swipperman, when it does go live, can you personally take a look at any deposit that comes in (of any type) and check it looks okay? If it doesn't, let Mark & I know.

**Comment by swipperman** [comment]
Can do!

**Comment by markmacgillivray** [comment]
Just keeping you up to date: everything is ready to move this to live, I've been checking through yesterday and today, was going to put it out today but haven't got it done yet. So tomorrow (my time) it will go onto the live service and I'll let you know of that. Then, once I've checked that's all behaving as it should, I'll switch over the URL routing so that live deposits will start going to the new system. I'll also update this issue when I do that, and that is when you'll start being able to check the live ones. That will be tomorrow too.

Further things to do outside of this issue, but which I've also been checking on to avoid later unexpected complexities, will be moving the user configs for those users who have deposit configs, and moving the old deposit records. I've got some more info from reviewing that data, but I'll make a separate issue for that data move, as it doesn't hold this back.

**Comment by markmacgillivray** [comment]
Hey, really sorry, I've been delayed today and will need to complete this tomorrow. Plan stands as previous comment, +1

**Comment by markmacgillivray** [comment]
Code on the live service has been updated and I've been keeping an eye on it for an hour or so. All looks well.

I'll update this issue again soon when I switch the URL routing so new live deposits start going into the new system.

**Comment by markmacgillivray** [comment]
The URL routing has been changed, any new live deposits should be handled by the new API. Let me know if any issues, I can switch the routing back if so.

**Comment by swipperman** [comment]
Fantastic! I'm going to send out a few [follow ups](https://github.com/oaworks/internal-planning/issues/236) today - I'll report back if I see any issues.

**Comment by JosephMcArthur** [comment]
Thanks @markmacgillivray, can we close this?

**Comment by swipperman** [comment]
Don't know if this is helpful or not, but I tried to deposit an AAM through a Gates link to SYP. It didn't work the first time ("Sorry we're unable to deposit your paper - please contact...") but did deposit successfully on the second (https://zenodo.org/record/6306117).

**Comment by JosephMcArthur** [comment]
hmmm, can you try to replicate that on dev? Do you know how to get to dev?
It's "dev." any address.

Joseph McArthur
Assistant Director: Right to Research Coalition <http://righttoresearch.org>
Director / Founder: Open Access Button <http://openaccessbutton.org>
Twitter: @Mcarthur_Joe <https://twitter.com/Mcarthur_Joe>
Skype: joseph_mcarthur

**Comment by swipperman** [comment]
I tried to deposit it twice through https://dev.shareyourpaper.org/funders/gates/10.1002/adtp.202100095?email=prausnitz@gatech.edu - once with the file that successfully deposited and a second time with the original downloaded file (I downloaded it twice, so the once that deposited just had a (1) at the end of the file name).

It successfully deposited both times: https://sandbox.zenodo.org/record/1023690, https://sandbox.zenodo.org/record/1023694

Sorry, I meant to take a screenshot the time it didn't deposit but accidentally hit the back button...

**Comment by JosephMcArthur** [comment]
So, you couldn't replicate what you saw the first time –– yes? In that case, it's likely something we can chalk up to internet/computer/people weirdness unless we see something like it again.

Another quick question for you. Did your first attempt on live come from a "review deposit" email?

**Comment by swipperman** [comment]
The link I used: https://shareyourpaper.org/funders/gates/10.1002/adtp.202100095?email=prausnitz@gatech.edu

**Comment by JosephMcArthur** [comment]
Sure does, thank you! I wondered if the special review link from the old system may not have worked in the new system (which would've been fine / maybe expected!). But, again, maybe just some weirdness. Thanks, this was a great report!

**Comment by markmacgillivray** [comment]
If you're both happy then yes this can be closed

**Comment by JosephMcArthur** [comment]
Thanks! @swipperman don't hesitate at all to reopen the moment you spot an issue.
